src/visual_editor/lib/hooks/useEditMode.ts

```diff
@@ -273,26 +288,118 @@ const useEditMode: UseEditModeHook = ({
   // upon every DOM mutation, we revert all changes and replay them to ensure
   // that the DOM is in the correct state
   const mutateRevert = useRef<(() => void) | null>(null);
+  
+  const runMutations = (domMutations?: VisualEditorVariation["domMutations"]) => {
+     // run reverts if they exist
+      if(!isGlobalObserverPaused()){
+        if (mutateRevert?.current) mutateRevert.current();
+      }
+      const revertCallbacks: Array<() => void> = [];
+      const stopCallbacks: Array<() => void> = [];
+      domMutations?.forEach((mutation) => {
+        const controller = mutate.declarative(mutation);
+        revertCallbacks.push(controller.revert);
+      });
+      if(!isGlobalObserverPaused()){
+        mutateRevert.current = () => {
+          revertCallbacks.reverse().forEach((c) => c());
+        }
+      }
+    }
+
   useEffect(() => {
-    // run reverts if they exist
-    if (mutateRevert?.current) mutateRevert.current();
 
-    const revertCallbacks: Array<() => void> = [];
+    runMutations(variation?.domMutations);
+  }, [variation]);
 
-    variation?.domMutations.forEach((mutation) => {
-      const controller = mutate.declarative(mutation);
-      revertCallbacks.push(controller.revert);
-    });
+  const hasTextInChildren = (element: Element) => {
+        for (let child of element.children) {
+            // Trim the child element's text content
+            const childText = child?.textContent?.trim();
+            if (childText || hasTextInChildren(child)) {
+                return true;
+            }
+        return false;
+    }
+  }
+
+
+  const resetAndStopInlineEditing = () => {
+    resumeGlobalObserver();
+
+    if(variation?.domMutations.length === 0 && elementUnderEdit){
+      elementUnderEdit.innerHTML = elementUnderEditCopy;
+    }
+
+    runMutations(variation?.domMutations);
+    
+    if (!elementUnderEdit) return;
+    elementUnderEdit.removeAttribute("contenteditable");
+    elementUnderEdit.removeEventListener("keydown", ()=>{});
```

Comment:
I'm pretty sure all of these `removeEventListener` calls are not doing anything.  The listener function must be passed in by reference, not declared inline like this.